Typo3 13 compatibility

hojalatheef opened the pull request.

hojalatheef committed:
[TASK] Updated CI Environment values

hojalatheef committed:
[TASK] Compatibility fixes for TYPO3 13 LTS

hojalatheef committed:
[TASK] Support for CI php 8.4

hojalatheef committed:
[TASK] Composer configuration update

hojalatheef committed:
[TASK] Fixed CGL Issues

hojalatheef committed:
[TASK] General compatibility fixes

hojalatheef committed:
[TASK] Replaced Hook with EventListener after ContentObjectRenderer is ready

hojalatheef committed:
[TASK] CGL Fixes

hojalatheef committed:
[TASK] Fixed UTC

hojalatheef committed:
[TASK] Formatting issues

hojalatheef committed:
[TASK] Removed unused function

sfroemkenjw commented on Classes/Command/CacheQueryCommand.php:
```diff
@@ -63,7 +63,7 @@ protected function execute(InputInterface $input, OutputInterface $output): int
 
         $cache = $this->getCache($input->getArgument('cacheIdentifier'));
-        if ($cache === null) {
+        if (!$cache instanceof FrontendInterface) {
```
nice

sfroemkenjw commented on Classes/Command/StatusReportCommand.php:
```diff
@@ -165,7 +167,7 @@ protected function getContentOfRobotsTxt(Site $site): string
             $base = rtrim((string)$site->getBase(), '/') . '/';
             $content = @file_get_contents($base . 'robots.txt');
-            $content = $content ?? '';
+            $content ??= '';
```
`file_get_contents` can only return `string` or `false` on failor, but never `null`. So don't check with `isset` here. If `false` it will be converted to `"0"`, that should be prevented. If `false` please return empty string. That way you don't need the additional check against `"0"` anymore.

sfroemkenjw commented on Classes/Controller/SolrController.php:
```diff
@@ -31,9 +32,7 @@
-/**
- * SolrController for backend module
- */
+#[AsController]
```
AsController is needed for non extbase Controllers to mark them `public:true` automatically. As AbstractController extends Extbase ActionController this PHP attribute is not needed, as there is a DI CompilerPass to mark it public.

sfroemkenjw commented on Classes/Controller/ToolsController.php:
```diff
@@ -12,19 +12,23 @@
-/**
- * ToolsController for backend module
- */
+#[AsController]
```
Not needed for Extbase controllers

sfroemkenjw commented on Classes/Controller/ToolsController.php:
```diff
@@ -12,19 +12,23 @@
+#[AsController]
 class ToolsController extends AbstractController
 {
+    public function __construct(
```
The AbstractController already includes this. Please remove.

sfroemkenjw commented on Classes/EventListener/AfterContentObjectRendererInitializedEventListener.php:
```diff
@@ -0,0 +1,54 @@
+)]
+final class AfterContentObjectRendererInitializedEventListener
+{
+    public function __construct(protected readonly ExtensionConfiguration $extensionConfiguration) {}
```
Class is `final`. Noone can overwrite anything. So please declare property as `private`

sfroemkenjw commented on Classes/EventListener/AfterContentObjectRendererInitializedEventListener.php:
```diff
@@ -0,0 +1,54 @@
+        $contentObjectRenderer = $event->getContentObjectRenderer();
+
+        if (
+            isset($contentObjectRenderer->parentRecord['data'][$contentObjectRenderer->currentValKey], $contentObjectRenderer->data['colPos']) && is_array(
+                $contentObjectRenderer->parentRecord,
+            ) && $contentObjectRenderer->parentRecord !== [] && ($this->getConfiguration(
+            )['typo3TransferTypoScriptCurrent'] ?? false) && $contentObjectRenderer->data['colPos'] !== 0
```
```suggestion
            isset($contentObjectRenderer->parentRecord['data'][$contentObjectRenderer->currentValKey], $contentObjectRenderer->data['colPos'])
                && is_array($contentObjectRenderer->parentRecord)
                && $contentObjectRenderer->parentRecord !== []
                && ($this->getConfiguration()['typo3TransferTypoScriptCurrent'] ?? false)
                && $contentObjectRenderer->data['colPos'] !== 0
```

sfroemkenjw commented on Classes/EventListener/ReduceCategoryTreeToPageTree.php:
```diff
@@ -102,11 +102,7 @@ protected function getPageUid(): int
-                } else {
-                    $pid = (int)$record['pid'];
-                }
+                $pid = empty($record) ? 0 : (int)$record['pid'];
```
Ihh "empty". `getRecordWSOL` can only return array and null.

sfroemkenjw commented on Classes/Hooks/CachingFrameworkLoggerHook.php:
```diff
@@ -68,20 +68,16 @@ public function analyze(array $parameters, VariableFrontend $frontend): void
-                        1720607181,
-                    );
-                }
+            // throw exception in FE only or also in BE, if fe_only is disabled
+            if ($cacheExpressionRecord['is_exception'] && ($cacheExpressionRecord['exception_fe_only'] === 0
+            || (
+                $cacheExpressionRecord['exception_fe_only'] === 1
+                && ApplicationType::fromRequest($this->getServerRequest())->isFrontend()
+            ))) {
```
Update to, or move into own method for better reading:

```suggestion
            // throw exception in FE only or also in BE, if fe_only is disabled
            if ($cacheExpressionRecord['is_exception']
                && (
                    $cacheExpressionRecord['exception_fe_only'] === 0
                    || (
                        $cacheExpressionRecord['exception_fe_only'] === 1
                        && ApplicationType::fromRequest($this->getServerRequest())->isFrontend()
                    )
                )
            ) {
```

sfroemkenjw commented on Classes/Hooks/ModifyElementInformationHook.php:
```diff
@@ -111,7 +115,7 @@ protected function init(ServerRequestInterface $request): void
         $queryParams = $request->getQueryParams();
         $this->table = $queryParams['table'] ?? null;
-        $this->uid = $queryParams['uid'] ?? null;
+        $this->uid = (int)$queryParams['uid'] ?? null;
```
This will break, if `uid` does not exists

sfroemkenjw commented on Configuration/user.tsconfig:
If it is empty, can we remove it?

sfroemkenjw commented on Tests/Unit/Task/ExecuteQueryTaskTest.php:
```diff
@@ -107,12 +94,11 @@ public function executeWithMultipleQueriesWillReturnTrue(): void
-            )
-            ->willReturn($statementMock);
+            ->willReturnOnConsecutiveCalls(
+                1,  // Return value for the first call (success)
+                1,  // Return value for the second call (success)
+                1,   // Return value for the third call (success)
```
```suggestion
                1, // Return value for the first call (success)
                1, // Return value for the second call (success)
                1, // Return value for the third call (success)
```

hojalatheef committed:
Update Tests/Unit/Task/ExecuteQueryTaskTest.php

Co-authored-by: Stefan Frömken <123929835+sfroemkenjw@users.noreply.github.com>

hojalatheef committed:
Update Classes/Hooks/CachingFrameworkLoggerHook.php

Co-authored-by: Stefan Frömken <123929835+sfroemkenjw@users.noreply.github.com>

hojalatheef committed:
[TASK] CR Fixes after review

- Deleted Configuration/user.tsconfig
- Removed php attributes for controller which is not required
- Modified code blocks and conditions which are used with empty

hojalatheef committed:
[TASK] CGL Fixes
